src/IO/S3/ProxyConfigurationResolverAdapter.h

```diff
@@ -0,0 +1,33 @@
+#pragma once
+
+#include "config.h"
+
+#if USE_AWS_S3
+
+#include <IO/S3/ProxyConfiguration.h>
+#include <Common/ProxyConfigurationResolver.h>
+
+namespace DB::S3
+{
+
+/*
+ * Common/ProxyConfigurationResolver has been built to serve as a generic Proxy Configuration Resolver as an attempt to re-use it
+ * in other parts of the project (e.g, URL functions). Therefore, it defines its own API and data types. To keep this backwards compatible
+ * and simplify usage on S3 existing APIs, this adapter is used to convert from Common/ProxyConfiguration to ClientConfigurationPerRequest.
+ * */
+class ProxyConfigurationResolverAdapter : public ProxyConfiguration
+{
+public:
+    explicit ProxyConfigurationResolverAdapter(std::shared_ptr<ProxyConfigurationResolver> resolver_)
+        : resolver(resolver_) {}
+    /// Returns proxy configuration on each HTTP request.
+    ClientConfigurationPerRequest getConfiguration(const Aws::Http::HttpRequest & request) override;
+    void errorReport(const ClientConfigurationPerRequest & config) override;
```

Comment:
Either I'm missing something or it should have failed to compile. The base class, `ProxyConfiguration` is a trivial struct with no virtual methods, and there are no other base classes. So there's nothing to override, these methods are not virtual

Reply:
Also, can we get rid of `ProxyConfigurationResolverAdapter` at all?  As far as I understood, `ClientConfigurationPerRequest` to `ProxyConfiguration` and vice versa (and the only difference between them is `ANY`, but there's another comment about this). 

There are only two usages of `ProxyConfigurationProvider` with `ProxyConfigurationResolverAdapter` that look like
```
    auto proxy_configuration_resolver = S3::ProxyConfigurationProvider::get(protocol);
    auto per_request_configuration = [=] (const Aws::Http::HttpRequest & req) { return proxy_configuration_resolver->getConfiguration(req); };
    auto error_report = [=] (const ClientConfigurationPerRequest & req) { proxy_configuration_resolver->errorReport(req); };
```

Let's just add two simple methods like `ProxyConfiguration::toClientConfiguration()` and `ProxyConfiguration::fromClientConfiguration(...)`. AFAIU, it will allow to replace the adapter with simple code like:
```
    auto proxy_configuration_resolver = S3::ProxyConfigurationResolverProvider::get(protocol);
    auto per_request_configuration = [=] (const Aws::Http::HttpRequest & req) { return proxy_configuration_resolver->resolve().toClientConfiguration(); };
    auto error_report = [=] (const ClientConfigurationPerRequest & req) { proxy_configuration_resolver->errorReport(ProxyConfiguration::fromClientConfiguration(req)); };
```

Or we can even override the implicit conversion operator...

Reply:
> Also, can we get rid of `ProxyConfigurationResolverAdapter` at all? As far as I understood, `ClientConfigurationPerRequest` to `ProxyConfiguration` and vice versa (and the only difference between them is `ANY`, but there's another comment about this).
> 
> There are only two usages of `ProxyConfigurationProvider` with `ProxyConfigurationResolverAdapter` that look like
> 
> ```
>     auto proxy_configuration_resolver = S3::ProxyConfigurationProvider::get(protocol);
>     auto per_request_configuration = [=] (const Aws::Http::HttpRequest & req) { return proxy_configuration_resolver->getConfiguration(req); };
>     auto error_report = [=] (const ClientConfigurationPerRequest & req) { proxy_configuration_resolver->errorReport(req); };
> ```
> 
> Let's just add two simple methods like `ProxyConfiguration::toClientConfiguration()` and `ProxyConfiguration::fromClientConfiguration(...)`. AFAIU, it will allow to replace the adapter with simple code like:
> 
> ```
>     auto proxy_configuration_resolver = S3::ProxyConfigurationResolverProvider::get(protocol);
>     auto per_request_configuration = [=] (const Aws::Http::HttpRequest & req) { return proxy_configuration_resolver->resolve().toClientConfiguration(); };
>     auto error_report = [=] (const ClientConfigurationPerRequest & req) { proxy_configuration_resolver->errorReport(ProxyConfiguration::fromClientConfiguration(req)); };
> ```
> 
> Or we can even override the implicit conversion operator...

I can do something along these lines for the sake of the PR, but I honestly prefer to avoid throwing the "burden" of converting stuff and "knowing implementation details" to the client. As of now, for the client, it's seamless. It does not know there are indirection layers, conversions and where the proxy configuration is coming from. If that changes, client code will not have to change.

I also don't like the idea of having a `DB::ProxyConfiguration` constructor that takes `ClientConfiguration` or the other way around. This adds extra dependencies.

I usually code with the notion of "the less a class knows, the better". If URL functions need to have access to `DB::ProxyConfiguration`, it should only see that. With those auxiliary methods or that constructor, it'll "see" `DB::S3::ClientConfiguration`, a useless dependency for URL functions (or anything outside the S3 scope).

I do understand though it's hard to keep track of a big number of classes and it might be confusing because of the similar names. Sometimes namespaces are not enough for us to keep track of that.

Let me know what you think.

Reply:
> "knowing implementation details" to the client. As of now, for the client, it's seamless

The client is basically 6 lines of code in 2 places (3 lines in each). Updating these 6 lines is much simpler than introducing some abstract adapters (about 100 lines of code).

>  If that changes, client code will not have to change.

If that changes, we can handle changing 6 lines of code.

> This adds extra dependencies.

It's not a big problem (we have a monorepo and a static binary). Also, it's quite natural to have an easy way to convert "new" `ProxyConfiguration` to "old" `ClientConfiguration` and vice versa

> I usually code with the notion of "the less a class knows, the better"

I understand this, and it's a really good principle sometimes. But I also like "less code is better" and "simpler is better". And in this particular case, all these abstractions look to me a bit overengineered (or maybe like "premature optimization", but in terms of abstractions)

> With those auxiliary methods or that constructor, it'll "see" DB::S3::ClientConfiguration, a useless dependency for URL functions (or anything outside the S3 scope).

I don't think it's a big problem, especially if URL function will not explicitly use this stuff and will only include some headers containing it. But if you want to do it the right way, then the right way is probably different.

S3 is basically the same as URL with something extra on top of basic HTTP(S). S3 and URL should depend on some common http stuff. And proxies must be a part of that common http stuff, it's not something S3-specific. So the right way is to refactor the code in the S3 scope, and replace `DB::S3::ClientConfiguration` with something that can be used in both URL and S3

Reply:
ok, fair enough. I'll see what I can do about this refactor, worst case scenario I'll add the auxiliary functions

Reply:
Actually, it's not that simple. There is extra logic implemented in `DB::S3::ProxyConfigurationProvider`, this extra logic is required to keep backwards compatibility.

To integrate the old behavior / settings with the new behavior / settings, here's what's implemented:

1. Try to fetch from scoped settings (storage conf proxy specific).
2. If it can't find storage conf specific proxy settings, try to fetch from general settings.
3. If it can't find there, grab from env.

This is the job of `DB::S3::ProxyConfigurationProvider`.

If I manage to remove `ClientConfigurationPerRequest` and use `DB::ProxyConfiguration` only, then maybe it's ok to eliminate `DB::S3::ProxyConfigurationResolverAdapter` and `DB::ProxyConfigurationProvider` by adding a method a backwards compatible method in `DB::ProxyConfigurationResolverProvider`. Will look into it

Reply:
Removed `ClientConfigurationPerRequest`, `DB::S3::ProxyConfigurationProvider`, `DB::S3::ProxyConfiguration` and `DB::S3::ProxyConfigurationResolverAdapter`.

Can you double check it's ok now?

Reply:
Thanks, will review it tomorrow